### PR TITLE
refactor: schema detection takes up to a minute

### PR DESCRIPTION
**Pull request by IyadhKhalfallah**
---

- [x] This PR references an open issue
- [x] Test added
- [x] Benchmark: from 65s to less than 1s locally

### CODE REVIEW

File: now/common/detect_schema.py [hunk 1]
```diff
@@ -148,10 +148,37 @@ def set_field_names_from_docarray(user_input: UserInput, **kwargs):
         raise ValueError('DocumentArray does not exist or you do not have access to it')
 
 
-def identify_folder_structure(file_paths: List[str], separator: str) -> str:
+def identify_s3_folder_structure(bucket, folder_prefix) -> str:
```
Comment (azayz) [inline now/common/detect_schema.py:151]: use only one function `identify_folder_structure` but add argument `is_s3_dataset` to know which code to use
 also add type hints

File: now/common/detect_schema.py [hunk 2]
```diff
@@ -148,10 +148,37 @@ def set_field_names_from_docarray(user_input: UserInput, **kwargs):
...
     """This function identifies the folder structure.
     It works with a local file structure or a remote file structure.
 
+    :param bucket: s3 bucket
+    :param folder_prefix: prefix of the s3 folder
+    :return: if all the files are in the same folder then returns 'single_folder' else returns 'sub_folders'
+    :raises Exception: if a file exists in sub_folder structure
+    """
+
+    structure_identifier = list(bucket.objects.filter(Prefix=folder_prefix).limit(2))[
```
Comment (azayz) [inline now/common/detect_schema.py:161]: add comments to explain what are you trying to get here, it's hard to understand just by looking at the code

File: now/common/detect_schema.py [hunk 3]
```diff
@@ -225,21 +252,24 @@ def set_field_names_from_s3_bucket(user_input: UserInput, **kwargs):
...
-    )  # user has to provide the folder where folder structure begins
-
-    objects = list(bucket.objects.filter(Prefix=folder_prefix))
-    file_paths = [
-        obj.key
-        for obj in objects
-        if not obj.key.endswith('/') and not obj.key.split('/')[-1].startswith('.')
-    ]
-    folder_structure = identify_folder_structure(file_paths, '/')
+    bucket, folder_prefix = get_s3_bucket_and_folder_prefix(user_input)
```
Comment (azayz) [inline now/common/detect_schema.py:255]: type hints

File: now/common/detect_schema.py [hunk 4]
```diff
@@ -225,21 +252,24 @@ def set_field_names_from_s3_bucket(user_input: UserInput, **kwargs):
...
+    # user has to provide the folder where folder structure begins
+    try:
+        first_file = list(bucket.objects.filter(Prefix=folder_prefix).limit(2))[1]
```
Comment (azayz) [inline now/common/detect_schema.py:258]: same comment as above

**Review by azayz — CHANGES_REQUESTED** [review]

**Review by azayz — APPROVED** [review]
LGTM after CI green, would love to see a small benchmark of the speed up

File: now/common/detect_schema.py [hunk 5]
```diff
@@ -44,10 +44,10 @@
~                search_fields_modalities[field_name] = MODALITIES_MAPPING[modality]
~                break
~            elif field_name == 'uri' and field_value.split('.')[-1] in file_types:
~                search_fields_modalities[field_name] = MODALITIES_MAPPING[modality]
~                break
```
Comment (joschkabraun) [inline now/common/detect_schema.py]: Let's remove this function if it's only used for s3 and is so short

File: now/common/detect_schema.py [hunk 6]
```diff
@@ -145,31 +145,18 @@
~            user_input.filter_fields_modalities,
~        ) = _extract_field_candidates_docarray(response)
~    else:
         raise ValueError('DocumentArray does not exist or you do not have access to it')
 
 
-def identify_folder_structure(file_paths: List[str], separator: str) -> str:
+def identify_folder_structure(first_file: str, folder_prefix: str) -> str:
```
Comment (joschkabraun) [inline now/common/detect_schema.py:151]: I would add the dataset path in quotation marks; that helps with understanding the error.

File: now/common/detect_schema.py [hunk 7]
```diff
@@ -145,31 +145,18 @@
~            user_input.filter_fields_modalities,
~        ) = _extract_field_candidates_docarray(response)
```
Comment (joschkabraun) [inline now/common/detect_schema.py]: what is `i` used for? I think we need to increment `i` below?

File: now/common/detect_schema.py [hunk 8]
```diff
@@ -145,31 +145,18 @@
...
~    else:
         raise ValueError('DocumentArray does not exist or you do not have access to it')
 
 
-def identify_folder_structure(file_paths: List[str], separator: str) -> str:
+def identify_folder_structure(first_file: str, folder_prefix: str) -> str:
     """This function identifies the folder structure.
-    It works with a local file structure or a remote file structure.
 
-    :param file_paths: list of file paths
```
Comment (joschkabraun) [inline now/common/detect_schema.py:155]: we can limit this to the first few hundred fails maybe?

File: now/data_loading/data_loading.py [hunk 9]
```diff
@@ -253,15 +255,24 @@ def _list_files_from_s3_bucket(
...
+    try:
+        # gets the first file in an s3 bucket, index 0 is reserved for the root folder name
+        first_file = list(bucket.objects.filter(Prefix=folder_prefix).limit(2))[1].key
+        i = 2
+        while first_file.split('/')[-1].startswith('.'):
+            first_file = list(bucket.objects.filter(Prefix=folder_prefix).limit(i + 1))[
+                i
+            ].key
+    except Exception as e:
+        raise Exception(f'Empty folder, data is missing.')
```
Comment (joschkabraun) [inline now/data_loading/data_loading.py:267]: this code looks familiar, let's wrap it in a function

**Review by joschkabraun — CHANGES_REQUESTED** [review]
Nice speed-up in detection!

File: now/common/detect_schema.py [hunk 10]
```diff
@@ -225,21 +212,30 @@ def set_field_names_from_s3_bucket(user_input: UserInput, **kwargs):
...
+    try:
+        # gets the first file in an s3 bucket, index 0 is reserved for the root folder name
+        first_file = list(bucket.objects.filter(Prefix=folder_prefix).limit(2))[1].key
+        i = 2
+        while first_file.split('/')[-1].startswith('.'):
+            first_file = list(bucket.objects.filter(Prefix=folder_prefix).limit(i + 1))[
+                i
+            ].key
+    except Exception as e:
+        raise Exception(f'Empty folder, data is missing.')
```
Comment (joschkabraun) [inline now/common/detect_schema.py:226]: I would add the dataset path in quotation marks; that helps with understanding the error.

File: now/common/detect_schema.py [hunk 11]
```diff
@@ -225,21 +212,30 @@ def set_field_names_from_s3_bucket(user_input: UserInput, **kwargs):
...
+
+    folder_structure = identify_folder_structure(first_file, folder_prefix)
     if folder_structure == 'single_folder':
+        objects = list(bucket.objects.filter(Prefix=folder_prefix))
```
Comment (joschkabraun) [inline now/common/detect_schema.py:230]: we can limit this to the first few hundred fails maybe?

File: now/common/detect_schema.py [hunk 12]
```diff
@@ -225,21 +212,30 @@ def set_field_names_from_s3_bucket(user_input: UserInput, **kwargs):
...
-        for obj in objects
-        if not obj.key.endswith('/') and not obj.key.split('/')[-1].startswith('.')
-    ]
-    folder_structure = identify_folder_structure(file_paths, '/')
+    bucket, folder_prefix = get_s3_bucket_and_folder_prefix(user_input)
+    # user has to provide the folder where folder structure begins
+    try:
+        # gets the first file in an s3 bucket, index 0 is reserved for the root folder name
+        first_file = list(bucket.objects.filter(Prefix=folder_prefix).limit(2))[1].key
+        i = 2
```
Comment (joschkabraun) [inline now/common/detect_schema.py:220]: I think we need to increment `i`, don't we?

File: now/common/detect_schema.py [hunk 13]
```diff
@@ -63,6 +63,21 @@ def _create_candidate_index_filter_fields(field_name_to_value):
     return index_fields_modalities, filter_field_modalities
 
 
+def get_first_file_in_folder_structure(bucket, folder_prefix, dataset_path):
```
Comment (joschkabraun) [inline now/common/detect_schema.py:66]: Let's call it `get_first_file_in_folder_structure_s3`

**Review by joschkabraun — APPROVED** [review]
LGTM!
Nice, speed-up.

**Pull request merged**